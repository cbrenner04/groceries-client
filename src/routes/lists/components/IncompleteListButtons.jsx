import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';
import axios from 'axios';

import { setUserInfo } from '../../../utils/auth';
import { Complete, Edit, Share, Trash } from '../../../components/ActionButtons';

function IncompleteListButtons(props) {
  const [currentUserPermissions, setCurrentUserPermissions] = useState('read');

  useEffect(() => {
    axios.get(`${process.env.REACT_APP_API_BASE}/lists/${props.list.id}/users_lists/${props.list.users_list_id}`, {
      headers: JSON.parse(sessionStorage.getItem('user')),
    }).then(({ data: { permissions }, headers }) => {
      setUserInfo(headers);
      setCurrentUserPermissions(permissions);
    }).catch(() => {
      // noop
    });
  }, [props.list.id, props.list.users_list_id]);

  return (
    <ButtonGroup className="float-right">
      <Complete
        handleClick={() => props.onListCompletion(props.list)}
        disabled={props.userId !== props.list.owner_id}
        style={{ opacity: props.userId !== props.list.owner_id ? 0.3 : 1 }}
        data-test-id="incomplete-list-complete"
      />
      <Share
        to={`lists/${props.list.id}/users_lists`}
        disabled={currentUserPermissions !== 'write'}
        style={{
          pointerEvents: currentUserPermissions !== 'write' ? 'none' : 'auto',
          opacity: currentUserPermissions !== 'write' ? 0.3 : 1,
        }}
        data-test-id="incomplete-list-share"
      />
      <Edit
        to={`/lists/${props.list.id}/edit`}
        disabled={props.userId !== props.list.owner_id}
        style={{
          pointerEvents: props.userId !== props.list.owner_id ? 'none' : 'auto',
          opacity: props.userId !== props.list.owner_id ? 0.3 : 1,
        }}
        data-test-id="incomplete-list-edit"
      />
      <Trash
        handleClick={() => props.onListDeletion(props.list)}
        disabled={props.userId !== props.list.owner_id}
        style={{ opacity: props.userId !== props.list.owner_id ? 0.3 : 1 }}
        data-test-id="incomplete-list-trash"
      />
    </ButtonGroup>
  );
}

IncompleteListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number,
  }).isRequired,
  onListCompletion: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
};

export default IncompleteListButtons;
