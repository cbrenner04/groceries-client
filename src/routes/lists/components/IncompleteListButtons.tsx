import React from 'react';
import { ButtonGroup } from 'react-bootstrap';

import { Complete, EditLink, Merge, Share, Trash } from 'components/ActionButtons';
import type { IList } from 'typings';

export interface IIncompleteListButtonsProps {
  userId: string;
  list: IList;
  onListCompletion: (list: IList) => void;
  onListDeletion: (list: IList) => void;
  currentUserPermissions: string;
  multiSelect: boolean;
  handleMerge: () => void;
  selectedLists: IList[];
  pending: boolean;
}

const IncompleteListButtons: React.FC<IIncompleteListButtonsProps> = (props): React.JSX.Element => {
  const userIsOwner = props.userId === props.list.owner_id;
  const multipleListsSelected = props.multiSelect && props.selectedLists.length > 1;
  const userCanShare = props.currentUserPermissions === 'write';

  return (
    <ButtonGroup className="float-end">
      <Complete
        handleClick={(): void => props.onListCompletion(props.list)}
        disabled={!userIsOwner || props.pending}
        classes={userIsOwner && !props.pending ? 'list-button-enabled' : 'list-button-disabled'}
        testID="incomplete-list-complete"
      />
      {!multipleListsSelected && (
        <Share
          to={`${props.list.id}/users_lists`}
          disabled={!userCanShare || props.pending}
          classes={userCanShare && !props.pending ? 'list-button-enabled' : 'list-button-disabled'}
          testID="incomplete-list-share"
        />
      )}
      {multipleListsSelected && (
        <Merge handleClick={props.handleMerge} testID="incomplete-list-merge" disabled={props.pending} />
      )}
      {!multipleListsSelected && (
        <EditLink
          to={`${props.list.id}/edit`}
          disabled={!userIsOwner || props.pending}
          classes={userIsOwner && !props.pending ? 'list-button-enabled' : 'list-button-disabled'}
          testID="incomplete-list-edit"
        />
      )}
      <Trash
        handleClick={(): void => props.onListDeletion(props.list)}
        testID="incomplete-list-trash"
        disabled={props.pending}
      />
    </ButtonGroup>
  );
};

export default IncompleteListButtons;
