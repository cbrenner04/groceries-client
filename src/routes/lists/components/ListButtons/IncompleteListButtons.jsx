import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Complete, EditLink, Merge, Share, Trash } from '../../../../components/ActionButtons';

function IncompleteListButtons(props) {
  const userIsOwner = props.userId === props.list.owner_id;
  const userCanShare = !props.multiSelect && props.currentUserPermissions === 'write';

  return (
    <ButtonGroup className="float-right">
      <Complete
        handleClick={() => props.onListCompletion(props.list)}
        disabled={!userIsOwner}
        style={{
          pointerEvents: userIsOwner ? 'auto' : 'none',
          opacity: userIsOwner ? 1 : 0.3,
        }}
        testID="incomplete-list-complete"
      />
      <Share
        to={`lists/${props.list.id}/users_lists`}
        disabled={!userCanShare}
        style={{
          pointerEvents: userCanShare ? 'auto' : 'none',
          opacity: userCanShare ? 1 : 0.3,
        }}
        testID="incomplete-list-share"
      />
      {props.multiSelect && <Merge handleClick={props.handleMerge} testID="incomplete-list-merge" />}
      {!props.multiSelect && (
        <EditLink
          to={`/lists/${props.list.id}/edit`}
          disabled={!userIsOwner}
          style={{
            pointerEvents: userIsOwner ? 'auto' : 'none',
            opacity: userIsOwner ? 1 : 0.3,
          }}
          testID="incomplete-list-edit"
        />
      )}
      <Trash handleClick={() => props.onListDeletion(props.list)} testID="incomplete-list-trash" />
    </ButtonGroup>
  );
}

IncompleteListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    owner_id: PropTypes.number.isRequired,
  }).isRequired,
  onListCompletion: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
  currentUserPermissions: PropTypes.string.isRequired,
  multiSelect: PropTypes.bool.isRequired,
  handleMerge: PropTypes.func.isRequired,
};

export default IncompleteListButtons;
