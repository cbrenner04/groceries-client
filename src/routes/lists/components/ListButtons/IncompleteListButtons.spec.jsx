import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import IncompleteListButtons from './IncompleteListButtons';

describe('IncompleteListButtons', () => {
  let props;
  const renderIncompleteListButtons = (props) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <IncompleteListButtons {...props} />
      </Router>,
    );
  };

  beforeEach(() => {
    props = {
      userId: 1,
      list: {
        id: 1,
        owner_id: 1,
      },
      onListCompletion: jest.fn(),
      onListDeletion: jest.fn(),
      currentUserPermissions: 'write',
      multiSelect: false,
      handleMerge: jest.fn(),
    };
  });

  it('complete and edit are disabled when user is not owner', () => {
    props.userId = 2;
    props.list.owner_id = 3;
    const { container, getByTestId } = renderIncompleteListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('incomplete-list-complete')).toBeDisabled();
    expect(getByTestId('incomplete-list-complete')).toHaveStyle({ opacity: 0.3 });
    expect(getByTestId('incomplete-list-edit')).toHaveAttribute('disabled', '');
    expect(getByTestId('incomplete-list-edit')).toHaveStyle({ opacity: 0.3, pointerEvents: 'none' });
  });

  it('complete and edit are enabled when user is owner', () => {
    props.userId = 1;
    props.list.owner_id = 1;
    const { container, getByTestId } = renderIncompleteListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('incomplete-list-complete')).toBeEnabled();
    expect(getByTestId('incomplete-list-complete')).toHaveStyle({ opacity: 1 });
    expect(getByTestId('incomplete-list-edit')).not.toHaveAttribute('disabled', '');
    expect(getByTestId('incomplete-list-edit')).toHaveStyle({ opacity: 1, pointerEvents: 'auto' });
  });

  it('share is disabled when user does not have write permissions', () => {
    props.currentUserPermissions = 'read';
    const { container, getByTestId } = renderIncompleteListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('incomplete-list-share')).toHaveAttribute('disabled', '');
    expect(getByTestId('incomplete-list-share')).toHaveStyle({ opacity: 0.3, pointerEvents: 'none' });
  });

  it('share is disabled when multiSelect', () => {
    props.multiSelect = true;
    const { container, getByTestId } = renderIncompleteListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('incomplete-list-share')).toHaveAttribute('disabled', '');
    expect(getByTestId('incomplete-list-share')).toHaveStyle({ opacity: 0.3, pointerEvents: 'none' });
  });

  it('share is enabled when user has write permissions and not multiSelect', () => {
    props.currentUserPermissions = 'write';
    props.multiSelect = false;
    const { container, getByTestId } = renderIncompleteListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('incomplete-list-share')).not.toHaveAttribute('disabled', '');
    expect(getByTestId('incomplete-list-share')).toHaveStyle({ opacity: 1, pointerEvents: 'auto' });
  });

  it('calls props.onListCompletion when complete is clicked', () => {
    const { getByTestId } = renderIncompleteListButtons(props);

    fireEvent.click(getByTestId('incomplete-list-complete'));

    expect(props.onListCompletion).toHaveBeenCalledWith(props.list);
  });

  it('calls props.onListDeletion when delete is clicked', () => {
    const { getByTestId } = renderIncompleteListButtons(props);

    fireEvent.click(getByTestId('incomplete-list-trash'));

    expect(props.onListDeletion).toHaveBeenCalledWith(props.list);
  });
});
