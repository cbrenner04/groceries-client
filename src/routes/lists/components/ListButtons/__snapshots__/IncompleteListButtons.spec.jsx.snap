// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncompleteListButtons complete and edit are disabled when user is not owner 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      disabled=""
      style="pointer-events: none; opacity: 0.3;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-share"
      href="/lists/1/users_lists"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </a>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-edit"
      disabled=""
      href="/lists/1/edit"
      style="pointer-events: none; opacity: 0.3;"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </a>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`IncompleteListButtons complete and edit are enabled when user is owner 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      style="pointer-events: auto; opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-share"
      href="/lists/1/users_lists"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </a>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-edit"
      href="/lists/1/edit"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </a>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`IncompleteListButtons share is disabled when multiSelect 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      style="pointer-events: auto; opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-share"
      disabled=""
      href="/lists/1/users_lists"
      style="pointer-events: none; opacity: 0.3;"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </a>
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-merge"
      type="button"
    >
      <i
        class="fa fa-compress-alt fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`IncompleteListButtons share is disabled when user does not have write permissions 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      style="pointer-events: auto; opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-share"
      disabled=""
      href="/lists/1/users_lists"
      style="pointer-events: none; opacity: 0.3;"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </a>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-edit"
      href="/lists/1/edit"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </a>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`IncompleteListButtons share is enabled when user has write permissions and not multiSelect 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      style="pointer-events: auto; opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-share"
      href="/lists/1/users_lists"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </a>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-edit"
      href="/lists/1/edit"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </a>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;
